### DISCUSSION TITLE
Retry doesn't seem to work with queryClient.fetchQuery

### DISCUSSION THREAD

**Discussion by Yuvasee** [TanStack/query#3558] [Q&A] [answered]
In my app I use mostly useQuery hook, but in few places queryClient.fetchQuery is needed. I use the same retry options and queryFn in all cases. I noticed that while with useQuery hook retry works just fine, fetchQuery doesn't catch errors in queryFn, so retry doesn't happen.

According to the docs I expect retry to just work in queryClient.fetchQuery. What do I misunderstand? Thanks.

**Comment by TkDodo** [comment, marked as answer]
retry should happen because it's an internal thing. There is an explicit code check though:

https://github.com/tannerlinsley/react-query/blob/f53d8b18c0396dee1052b3bbdbbf3e68b589dc99/src/core/queryClient.ts#L361-L364

that references this issue:
https://github.com/tannerlinsley/react-query/issues/652

I'm not really sure if the implementation is correct. Judging from the comments, it seems like we _wanted_ to turn off default retries for `prefetchQuery`, but instead did it on `fetchQuery` - which also works for prefetching, because `prefetchQuery` just calls `fetchQuery` internally.

Looking at the PR, which was made for v2, this used be in the prefetching function, but potentially regressed in the v3 rewrite:

https://github.com/tannerlinsley/react-query/pull/760

I would accept a PR that moves this logic to the prefetchQuery function with the necessary tests. Would you like to contribute?

In the meantime, passing retry explicitly should also work:

```
queryClient.fetchQuery(key, fn, { retry: 3 })
```

**Reply by Yuvasee** [reply]
Thank you for the quick answer. After further investigation I realized that the retry logic works just as expected and retries do happen. I was confused by the fact that `fetchQuery` throws after all attempts failed and useQuery in my case just returns the error status.

**Reply by TkDodo** [reply]
It does throw between each retry? Don't think it should. Can you show that in a codesandbox?

**Reply by overyew** [reply]
I was having this issue where the retry keeps raising exceptions and i was not able to find any reference of this feature anywhere. I can assure you it is still happening.
i have a wrapper that uses the fetchQuery method with a retry set to 3 and it is raising 4 exception for a call that is bound to fail.

**Reply by TkDodo** [reply]
then show a minimal, standalone reproduction please.
